### ISSUE TITLE
Console: Semantically unstable OutputFormatter::escape() and ::formatAndWrap(); performance issues

### ISSUE THREAD

**Issue by exteon** [symfony/symfony#34065]
**Symfony version(s) affected**: all

**Description**  
Calling OutputFormatter::escape() on `A\<B>` or `A<B>` returns the same string: `A\<B>`. This is inconsistent for an escape/unescape function, as well as producing unreliable output (granted, it's an edge case, but still).

**How to reproduce**  
```php
echo OutputFormatter::escape('A\\<B>');
echo OutputFormatter::escape('A<B>');
```

**Possible Solution**  
Always duplicate `\`s before `<` on escape;  then use `"#(\\\\)*<($tagRegex) | /($tagRegex)?)>#` for matching; if `$1` that matched has an even number of characters, it's a tag; if odd, it's an escape tag. Then deduplicate the leading `\`s.

**Of (lesser?) consequence**  
Using `$output.=...` in a (possibly huge) buffering context is a bit of a don't, performance-wise. `$fragments[]=...;` then finally and only once `implode('',$fragments)` is much faster, reason being `.` operator always creates a copy of the leftmost operand. Test for instance the performance of:
```
$outputFormatter->formatAndWrap(str_repeat('A<B>C</>',100000));
```
You will see it's O(n^2). 

**Relevance**
Probably small in everyday life, but we were thinking of building upon the `Console\Output` to create  a streamable multitarget output dispatcher (i.e. stream the same initial output to log files as text, to a browser-based event watcher as HTML using a HTML formatter, to console using the ANSI formatter etc).

**Note about the trailing \ **
Since the intermediary medium of an `escape()` / `unescape()` sequence is formally an encoding, and escape sequences are characters in that encoding (and therefore atomic), it is reasonable to expect a trailing `\` is not part of the next escape sequence, which could happen only if `formatAndWrap()` was replayed input from a block bytestream. To be extra preventive, one could save the trailing `\` in state (`OutputFormatter` is stateful anyway via `$styleStack`), but that would require a `flush()` operation at the end. On the other hand, the presence of `$styleStack` kindof warrants the existence of a `flush()` anyway that would close the open styles on the stream.

**Comment by carsonbot** [comment]
Thank you for this issue.
There has not been a lot of activity here for a while. Has this been resolved?

**Comment by carsonbot** [comment]
Hello? This issue is about to be closed if nobody replies.

**Comment by carsonbot** [comment]
Hey,

I didn't hear anything so I'm going to close it. Feel free to comment if this is still relevant, I can always reopen!
